File: crypto_screener/collector.py
from __future__ import annotations

import os
import re
import time
from typing import Any

from .coingecko import CoinGeckoClient
from .coinglass import CoinGlassClient
from .derivatives import derivatives_snapshot
from .providers import ProviderError
from .quality import apply_data_quality
from .scoring import funding_annualized_pct, to_float
from .technicals import technical_snapshot


def collect_market(config: dict[str, Any]) -> dict[str, Any]:
    status: dict[str, Any] = {}
    rows = collect_coinglass_futures(config, status)
    market_context = collect_coingecko_context(config, status)
    status["data_quality"] = apply_data_quality(rows, config)
    return {
        "rows": rows,
        "market_context": market_context,
        "provider_status": status,
    }


def collect_coinglass_futures(config: dict[str, Any], status: dict[str, Any] | None = None) -> list[dict[str, Any]]:
    provider_cfg = config.get("providers", {}).get("coinglass", {})
    universe_cfg = config.get("universe", {})
    if not provider_cfg.get("enabled", True):
        raise ProviderError("CoinGlass provider is required for futures collection")

    api_key_env = provider_cfg.get("api_key_env", "COINGLASS_API_KEY")
    api_key = os.environ.get(api_key_env, "")
    if not api_key:
        raise ProviderError(f"{api_key_env} is required for CoinGlass-only futures collection")

    client = CoinGlassClient(
        api_key=api_key,
        base_url=provider_cfg.get("base_url", "https://open-api-v4.coinglass.com"),
        timeout_seconds=float(provider_cfg.get("request_timeout_seconds", 12)),
    )
    exchanges = set(provider_cfg.get("exchanges", []))
    request_delay = float(provider_cfg.get("request_delay_seconds", 2.1))
    top_symbols = int(universe_cfg.get("top_symbols_by_volume", 80))
    candidate_limit = int(provider_cfg.get("candidate_symbols", top_symbols))
    min_volume = float(universe_cfg.get("min_quote_volume_usd", 20_000_000))
    quote_asset = str(universe_cfg.get("quote_asset", "USDT"))
    min_exchange_count = int(provider_cfg.get("min_exchange_count", 2))
    excluded_bases = {str(item).upper() for item in universe_cfg.get("exclude_base_assets", [])}
    core_symbols = [str(item).upper() for item in config.get("report", {}).get("core_symbols", ["BTC", "ETH", "SOL"])]

    supported_pairs = client.supported_exchange_pairs()
    candidate_stats = _coinglass_candidate_stats(
        supported_pairs=supported_pairs,
        exchanges=exchanges,
        quote_asset=quote_asset,
        min_exchange_count=min_exchange_count,
        excluded_bases=excluded_bases,
    )
    candidates = _rank_coinglass_candidates(candidate_stats, core_symbols, candidate_limit)

    rows: list[dict[str, Any]] = []
    errors: list[str] = []
    for symbol in candidates:
        try:
            pairs = client.futures_pairs_markets(symbol)
            aggregate = _aggregate_coinglass_pairs(pairs, exchanges, candidate_stats.get(symbol, {}), quote_asset)
            if aggregate and (aggregate.get("quote_volume_usd") or 0.0) >= min_volume:
                rows.append(aggregate)
        except ProviderError as exc:
            errors.append(f"{symbol}: {exc}")
        finally:
            if request_delay > 0:
                time.sleep(request_delay)

    rows.sort(key=lambda row: row.get("quote_volume_usd") or 0.0, reverse=True)
    rows = rows[:top_symbols]
    _append_coinglass_technicals(rows, client, provider_cfg, status)
    _append_coinglass_derivatives_history(rows, client, provider_cfg, status)

    if status is not None:
        status["coinglass"] = {
            "status": "ok" if rows else "error",
            "rows": len(rows),
            "candidate_symbols": len(candidates),
            "supported_symbols": len(candidate_stats),
            "errors": errors[:5],
            "note": "CoinGlass futures pairs-markets primary data",
        }
    return rows


def _append_coinglass_technicals(
    rows: list[dict[str, Any]],
    client: CoinGlassClient,
    provider_cfg: dict[str, Any],
    status: dict[str, Any] | None,
) -> None:
    technical_cfg = provider_cfg.get("technical_indicators", {})
    if not technical_cfg.get("enabled", True):
        if status is not None:
            status["technicals"] = {"status": "disabled"}
        return

    interval = str(technical_cfg.get("interval", "4h"))
    limit = int(technical_cfg.get("limit", 220))
    max_symbols = int(technical_cfg.get("max_symbols", 40))
    request_delay = float(technical_cfg.get("request_delay_seconds", provider_cfg.get("request_delay_seconds", 2.1)))
    enriched = 0
    errors: list[str] = []

    for row in rows[:max_symbols]:
        exchange = str(row.get("primary_exchange") or "")
        contract_symbol = str(row.get("contract_symbol") or "")
        if not exchange or not contract_symbol:
            continue
        try:
            candles = client.price_history(exchange, contract_symbol, interval, limit)
            snapshot = technical_snapshot(candles, interval)
            if snapshot:
                row.update(snapshot)
                enriched += 1
        except ProviderError as exc:
            errors.append(f"{row.get('symbol', contract_symbol)}: {exc}")
        finally:
            if request_delay > 0:
                time.sleep(request_delay)

    if status is not None:
        status["technicals"] = {
            "status": "ok" if enriched else "error",
            "rows": enriched,
            "candidate_symbols": min(max_symbols, len(rows)),
            "interval": interval,
            "errors": errors[:5],
            "note": "CoinGlass futures price OHLC technical indicators",
        }


def _append_coinglass_derivatives_history(
    rows: list[dict[str, Any]],
    client: CoinGlassClient,
    provider_cfg: dict[str, Any],
    status: dict[str, Any] | None,
) -> None:
    history_cfg = provider_cfg.get("derivatives_history", {})
    if not history_cfg.get("enabled", True):
        if status is not None:
            status["derivatives_history"] = {"status": "disabled"}
        return

    interval = str(history_cfg.get("interval", "4h"))
    limit = int(history_cfg.get("limit", 220))
    max_symbols = int(history_cfg.get("max_symbols", 30))
    request_delay = float(history_cfg.get("request_delay_seconds", provider_cfg.get("request_delay_seconds", 2.1)))
    exchanges = [str(item) for item in provider_cfg.get("exchanges", [])]
    enriched = 0
    errors: list[str] = []

    for row in rows[:max_symbols]:
        symbol = str(row.get("symbol") or "")
        if not symbol:
            continue
        try:
            oi_history = client.open_interest_aggregated_history(symbol, interval, limit)
            _sleep_between_requests(request_delay)
            funding_history = client.funding_oi_weight_history(symbol, interval, limit)
            _sleep_between_requests(request_delay)
            liquidation_history = client.liquidation_aggregated_history(exchanges, symbol, interval, limit)
            _sleep_between_requests(request_delay)
            taker_history = client.aggregated_taker_buy_sell_history(exchanges, symbol, interval, limit)

            snapshot = derivatives_snapshot(
                oi_history=oi_history,
                funding_history=funding_history,
                liquidation_history=liquidation_history,
                taker_history=taker_history,
                interval=interval,
            )
            if snapshot:
                row.update(snapshot)
                enriched += 1
        except ProviderError as exc:
            errors.append(f"{symbol}: {exc}")
        finally:
            if request_delay > 0:
                time.sleep(request_delay)

    if status is not None:
        status["derivatives_history"] = {
            "status": "ok" if enriched else "error",
            "rows": enriched,
            "candidate_symbols": min(max_symbols, len(rows)),
            "interval": interval,
            "errors": errors[:5],
            "note": "CoinGlass historical OI/funding/liquidation/taker features",
        }


def collect_coingecko_context(config: dict[str, Any], status: dict[str, Any]) -> dict[str, Any]:
    provider_cfg = config.get("providers", {}).get("coingecko", {})
    if not provider_cfg.get("enabled", True):
        status["coingecko"] = {"status": "disabled"}
        return {}

    api_key = os.environ.get(provider_cfg.get("api_key_env", "COINGECKO_API_KEY"), "")
    client = CoinGeckoClient(
        base_url=provider_cfg.get("base_url", "https://api.coingecko.com/api/v3"),
        api_key=api_key or None,
        timeout_seconds=float(provider_cfg.get("request_timeout_seconds", 12)),
        retry_429=bool(provider_cfg.get("retry_429", True)),
        retry_429_initial_delay_seconds=float(provider_cfg.get("retry_429_initial_delay_seconds", 30)),
        retry_429_max_delay_seconds=float(provider_cfg.get("retry_429_max_delay_seconds", 300)),
        retry_429_jitter_seconds=float(provider_cfg.get("retry_429_jitter_seconds", 15)),
        retry_429_max_attempts=int(provider_cfg.get("retry_429_max_attempts", 0)),
    )
    context: dict[str, Any] = {}
    errors: list[str] = []
    try:
        global_data = client.global_data()
        context.update(_normalize_coingecko_global(global_data))
    except ProviderError as exc:
        errors.append(str(exc))

    try:
        categories = client.categories()
        limit = int(provider_cfg.get("categories_limit", 12))
        context["categories"] = _normalize_coingecko_categories(categories, limit)
    except ProviderError as exc:
        errors.append(str(exc))

    status["coingecko"] = {
        "status": "ok" if context else "error",
        "errors": errors[:5],
        "note": "global market and category context",
    }
    return context


def _sleep_between_requests(seconds: float) -> None:
    if seconds > 0:
        time.sleep(seconds)


def _coinglass_candidate_stats(
    supported_pairs: dict[str, list[dict[str, Any]]],
    exchanges: set[str],
    quote_asset: str,
    min_exchange_count: int,
    excluded_bases: set[str],
) -> dict[str, dict[str, Any]]:
    stats: dict[str, dict[str, Any]] = {}
    for exchange_name, pairs in supported_pairs.items():
        if exchanges and exchange_name not in exchanges:
            continue
        for pair in pairs:
            base_asset = str(pair.get("base_asset") or "").upper()
            if not base_asset or base_asset in excluded_bases:
                continue
            if not _quote_matches(pair, quote_asset):
                continue
            if not _is_likely_perpetual(pair):
                continue
            item = stats.setdefault(
                base_asset,
                {
                    "symbol": base_asset,
                    "exchanges": set(),
                    "instrument_count": 0,
                    "max_leverage": 0.0,
                },
            )
            item["exchanges"].add(exchange_name)
            item["instrument_count"] += 1
            item["max_leverage"] = max(item["max_leverage"], to_float(pair.get("max_leverage"), 0.0) or 0.0)

    return {
        symbol: item
        for symbol, item in stats.items()
        if len(item["exchanges"]) >= min_exchange_count
    }


def _rank_coinglass_candidates(
    candidate_stats: dict[str, dict[str, Any]],
    core_symbols: list[str],
    limit: int,
) -> list[str]:
    ranked = sorted(
        candidate_stats,
        key=lambda symbol: (
            len(candidate_stats[symbol].get("exchanges", [])),
            candidate_stats[symbol].get("instrument_count", 0),
            candidate_stats[symbol].get("max_leverage", 0.0),
            symbol,
        ),
        reverse=True,
    )
    ordered: list[str] = []
    for symbol in core_symbols + ranked:
        if symbol in candidate_stats and symbol not in ordered:
            ordered.append(symbol)
        if len(ordered) >= limit:
            break
    return ordered


def _aggregate_coinglass_pairs(
    pairs: list[dict[str, Any]],
    exchanges: set[str],
    symbol_stats: dict[str, Any],
    quote_asset: str,
) -> dict[str, Any]:
    filtered = [
        pair
        for pair in pairs
        if (not exchanges or pair.get("exchange_name") in exchanges)
        and _pair_symbol_matches_quote(pair, quote_asset)
    ]
    if not filtered:
        return {}

    primary = max(filtered, key=lambda pair: to_float(pair.get("volume_usd"), 0.0) or 0.0)
    symbol = _base_from_pair(primary)
    total_volume = sum(to_float(pair.get("volume_usd"), 0.0) or 0.0 for pair in filtered)
    total_oi = sum(to_float(pair.get("open_interest_usd"), 0.0) or 0.0 for pair in filtered)
    long_volume = sum(to_float(pair.get("long_volume_usd"), 0.0) or 0.0 for pair in filtered)
    short_volume = sum(to_float(pair.get("short_volume_usd"), 0.0) or 0.0 for pair in filtered)
    long_liq = sum(to_float(pair.get("long_liquidation_usd_24h"), 0.0) or 0.0 for pair in filtered)
    short_liq = sum(to_float(pair.get("short_liquidation_usd_24h"), 0.0) or 0.0 for pair in filtered)
    funding = _weighted_average(filtered, "funding_rate", "open_interest_usd")

    return {
        "symbol": symbol,
        "contract_symbol": primary.get("instrument_id") or f"{symbol}{quote_asset}",
        "base_asset": symbol,
        "quote_asset": quote_asset,
        "primary_exchange": primary.get("exchange_name"),
        "data_source": "coinglass",
        "price_usd": _weighted_average(filtered, "current_price", "volume_usd"),
        "index_price": _weighted_average(filtered, "index_price", "volume_usd"),
        "price_change_24h_pct": _weighted_average(filtered, "price_change_percent_24h", "volume_usd"),
        "quote_volume_usd": total_volume,
        "volume_change_percent_24h": _weighted_average(filtered, "volume_usd_change_percent_24h", "volume_usd"),
        "open_interest_usd": total_oi,
        "oi_change_24h_pct": _weighted_average(filtered, "open_interest_change_percent_24h", "open_interest_usd"),
        "funding_rate_pct": funding,
        "funding_annualized_pct": funding_annualized_pct(funding / 100.0) if funding is not None else None,
        "next_funding_time": primary.get("next_funding_time"),
        "long_volume_usd_24h": long_volume,
        "short_volume_usd_24h": short_volume,
        "long_short_ratio": (long_volume / short_volume) if short_volume > 0 else None,
        "long_liquidation_usd_24h": long_liq,
        "short_liquidation_usd_24h": short_liq,
        "open_interest_volume_ratio": (total_oi / total_volume) if total_volume > 0 else None,
        "coinglass_exchange_count": len({pair.get("exchange_name") for pair in filtered}),
        "coinglass_instrument_count": symbol_stats.get("instrument_count"),
        "coinglass_supported_exchange_count": len(symbol_stats.get("exchanges", [])),
    }


def _quote_matches(pair: dict[str, Any], quote_asset: str) -> bool:
    return (
        str(pair.get("quote_asset") or "").upper() == quote_asset
        or str(pair.get("settlement_currency") or "").upper() == quote_asset
    )


def _pair_symbol_matches_quote(pair: dict[str, Any], quote_asset: str) -> bool:
    symbol = str(pair.get("symbol") or "").upper()
    instrument_id = str(pair.get("instrument_id") or "").upper()
    return symbol.endswith(f"/{quote_asset}") or quote_asset in instrument_id


def _is_likely_perpetual(pair: dict[str, Any]) -> bool:
    instrument_id = str(pair.get("instrument_id") or "")
    lowered = instrument_id.lower()
    if "perp" in lowered or "swap" in lowered:
        return True
    return re.search(r"[_-]\d{6,8}$", instrument_id) is None


def _base_from_pair(pair: dict[str, Any]) -> str:
    symbol = str(pair.get("symbol") or "")
    if "/" in symbol:
        return symbol.split("/", 1)[0].upper()
    instrument_id = str(pair.get("instrument_id") or "")
    return re.sub(r"[^A-Z0-9].*$", "", instrument_id.upper()).replace("USDT", "")


def _normalize_coingecko_global(global_data: dict[str, Any]) -> dict[str, Any]:
    total_market_cap = global_data.get("total_market_cap", {})
    market_cap_percentage = global_data.get("market_cap_percentage", {})
    return {
        "total_market_cap_usd": to_float(total_market_cap.get("usd")),
        "market_cap_change_24h_pct": to_float(global_data.get("market_cap_change_percentage_24h_usd")),
        "btc_dominance_pct": to_float(market_cap_percentage.get("btc")),
        "eth_dominance_pct": to_float(market_cap_percentage.get("eth")),
        "active_cryptocurrencies": global_data.get("active_cryptocurrencies"),
        "markets": global_data.get("markets"),
    }


def _normalize_coingecko_categories(categories: list[dict[str, Any]], limit: int) -> dict[str, list[dict[str, Any]]]:
    normalized = [
        {
            "id": item.get("id"),
            "name": item.get("name"),
            "market_cap_usd": to_float(item.get("market_cap")),
            "market_cap_change_24h_pct": to_float(item.get("market_cap_change_24h")),
            "volume_24h_usd": to_float(item.get("volume_24h")),
            "top_3_coins": item.get("top_3_coins", []),
        }
        for item in categories
        if to_float(item.get("market_cap_change_24h")) is not None
    ]
    leaders = sorted(normalized, key=lambda item: item["market_cap_change_24h_pct"], reverse=True)[:limit]
    laggards = sorted(normalized, key=lambda item: item["market_cap_change_24h_pct"])[:limit]
    return {"leaders": leaders, "laggards": laggards}


def _weighted_average(rows: list[dict[str, Any]], value_key: str, weight_key: str) -> float | None:
    weighted_sum = 0.0
    total_weight = 0.0
    for row in rows:
        value = to_float(row.get(value_key))
        weight = to_float(row.get(weight_key), 0.0) or 0.0
        if value is None or weight <= 0:
            continue
        weighted_sum += value * weight
        total_weight += weight
    return weighted_sum / total_weight if total_weight > 0 else None

File: apps/api/src/pipeline/collector.ts
import type { AppConfig } from '../config/index.js';
import type { CoinGeckoClient } from '../providers/coingecko.js';
import { CoinGeckoHttpClient } from '../providers/coingecko.js';
import type { CoinGlassClient, CoinGlassPair } from '../providers/coinglass.js';
import { CoinGlassHttpClient } from '../providers/coinglass.js';
import { ProviderError } from '../providers/errors.js';
import { sleep } from '../providers/http.js';
import {
  baseFromPair,
  isLikelyPerpetualPair,
  pairSymbolMatchesQuote,
  quoteMatches,
} from './coinglassPairs.js';
import type { ProviderStatus } from './enrichment.js';
import {
  appendCoinglassDerivativesHistory,
  appendCoinglassLongShortRatio,
  appendCoinglassTechnicals,
} from './enrichment.js';
import { applyDataQuality } from './quality.js';
import { fundingAnnualizedPct, toFloat, weightedAverage } from './scoring.js';
import type { Row } from './types.js';
import { asRecord } from './types.js';

export type { ProviderStatus } from './enrichment.js';

export interface CollectResult {
  rows: Row[];
  market_context: Record<string, unknown>;
  provider_status: ProviderStatus;
}

export interface CollectDeps {
  coinglassClient?: CoinGlassClient;
  coingeckoClient?: CoinGeckoClient;
}

export async function collectMarket(
  config: AppConfig,
  deps: CollectDeps = {},
): Promise<CollectResult> {
  const status: ProviderStatus = {};
  const rows = await collectCoinglassFutures(config, status, deps.coinglassClient);
  const marketContext = await collectCoingeckoContext(config, status, deps.coingeckoClient);
  status.data_quality = applyDataQuality(rows, config);
  return { rows, market_context: marketContext, provider_status: status };
}

export async function collectCoinglassFutures(
  config: AppConfig,
  status?: ProviderStatus,
  client?: CoinGlassClient,
): Promise<Row[]> {
  const providerCfg = config.providers.coinglass;
  const universeCfg = config.universe;
  if (!providerCfg.enabled) {
    throw new ProviderError('CoinGlass provider is required for futures collection');
  }

  const apiKeyEnv = providerCfg.api_key_env || 'COINGLASS_API_KEY';
  const apiKey = process.env[apiKeyEnv] ?? '';
  if (!client && !apiKey) {
    throw new ProviderError(`${apiKeyEnv} is required for CoinGlass-only futures collection`);
  }

  const coinglassClient =
    client ??
    new CoinGlassHttpClient({
      apiKey,
      baseUrl: providerCfg.base_url,
      timeoutSeconds: providerCfg.request_timeout_seconds,
    });

  const exchanges = new Set(providerCfg.exchanges);
  const requestDelay = providerCfg.request_delay_seconds;
  const topSymbols = universeCfg.top_symbols_by_volume;
  const candidateLimit = providerCfg.candidate_symbols || topSymbols;
  const minVolume = universeCfg.min_quote_volume_usd;
  const quoteAsset = universeCfg.quote_asset;
  const minExchangeCount = providerCfg.min_exchange_count;
  const excludedBases = new Set(universeCfg.exclude_base_assets.map((item) => item.toUpperCase()));
  const coreSymbols = config.report.core_symbols.map((item) => item.toUpperCase());

  const supportedPairs = await coinglassClient.supportedExchangePairs();
  const candidateStats = coinglassCandidateStats({
    supportedPairs,
    exchanges,
    quoteAsset,
    minExchangeCount,
    excludedBases,
  });
  const candidates = rankCoinglassCandidates(candidateStats, coreSymbols, candidateLimit);

  const rows: Row[] = [];
  const errors: string[] = [];
  for (const symbol of candidates) {
    try {
      const pairs = await coinglassClient.futuresPairsMarkets(symbol);
      const aggregate = aggregateCoinglassPairs(
        pairs,
        exchanges,
        candidateStats.get(symbol) ?? null,
        quoteAsset,
      );
      const quoteVolumeUsd = toFloat(aggregate?.quote_volume_usd, 0.0) ?? 0.0;
      if (aggregate && quoteVolumeUsd >= minVolume) {
        rows.push(aggregate);
      }
    } catch (error) {
      if (error instanceof ProviderError) {
        errors.push(`${symbol}: ${error.message}`);
      } else {
        throw error;
      }
    } finally {
      if (requestDelay > 0) {
        await sleep(requestDelay);
      }
    }
  }

  rows.sort(
    (a, b) => (toFloat(b.quote_volume_usd, 0.0) ?? 0.0) - (toFloat(a.quote_volume_usd, 0.0) ?? 0.0),
  );
  const topRows = rows.slice(0, topSymbols);
  await appendCoinglassTechnicals(topRows, coinglassClient, providerCfg, status);
  await appendCoinglassDerivativesHistory(topRows, coinglassClient, providerCfg, status);
  await appendCoinglassLongShortRatio(topRows, coinglassClient, providerCfg, status);

  if (status) {
    status.coinglass = {
      status: topRows.length ? 'ok' : 'error',
      rows: topRows.length,
      candidate_symbols: candidates.length,
      supported_symbols: candidateStats.size,
      errors: errors.slice(0, 5),
      note: 'CoinGlass futures pairs-markets primary data',
    };
  }
  return topRows;
}

export async function collectCoingeckoContext(
  config: AppConfig,
  status: ProviderStatus,
  client?: CoinGeckoClient,
): Promise<Record<string, unknown>> {
  const providerCfg = config.providers.coingecko;
  if (!providerCfg.enabled) {
    status.coingecko = { status: 'disabled' };
    return {};
  }

  const apiKey = process.env[providerCfg.api_key_env || 'COINGECKO_API_KEY'] ?? '';
  const coingeckoClient =
    client ??
    new CoinGeckoHttpClient({
      baseUrl: providerCfg.base_url,
      apiKey: apiKey || null,
      timeoutSeconds: providerCfg.request_timeout_seconds,
      retry429: providerCfg.retry_429,
      retry429InitialDelaySeconds: providerCfg.retry_429_initial_delay_seconds,
      retry429MaxDelaySeconds: providerCfg.retry_429_max_delay_seconds,
      retry429JitterSeconds: providerCfg.retry_429_jitter_seconds,
      retry429MaxAttempts: providerCfg.retry_429_max_attempts,
    });

  const context: Record<string, unknown> = {};
  const errors: string[] = [];
  try {
    const globalData = await coingeckoClient.globalData();
    Object.assign(context, normalizeCoingeckoGlobal(globalData));
  } catch (error) {
    if (error instanceof ProviderError) {
      errors.push(error.message);
    } else {
      throw error;
    }
  }

  try {
    const categories = await coingeckoClient.categories();
    context.categories = normalizeCoingeckoCategories(categories, providerCfg.categories_limit);
  } catch (error) {
    if (error instanceof ProviderError) {
      errors.push(error.message);
    } else {
      throw error;
    }
  }

  status.coingecko = {
    status: Object.keys(context).length ? 'ok' : 'error',
    errors: errors.slice(0, 5),
    note: 'global market and category context',
  };
  return context;
}

export interface CandidateStats {
  symbol: string;
  exchanges: Set<string>;
  instrumentCount: number;
  maxLeverage: number;
}

/** Exported for unit testing. */
export function coinglassCandidateStats(options: {
  supportedPairs: Record<string, CoinGlassPair[]>;
  exchanges: Set<string>;
  quoteAsset: string;
  minExchangeCount: number;
  excludedBases: Set<string>;
}): Map<string, CandidateStats> {
  const { supportedPairs, exchanges, quoteAsset, minExchangeCount, excludedBases } = options;
  const stats = new Map<string, CandidateStats>();

  for (const [exchangeName, pairs] of Object.entries(supportedPairs)) {
    if (exchanges.size > 0 && !exchanges.has(exchangeName)) {
      continue;
    }
    for (const pair of pairs) {
      const baseAsset = String(pair.base_asset ?? '').toUpperCase();
      if (!baseAsset || excludedBases.has(baseAsset)) {
        continue;
      }
      if (!quoteMatches(pair, quoteAsset)) {
        continue;
      }
      if (!isLikelyPerpetualPair(pair)) {
        continue;
      }
      let item = stats.get(baseAsset);
      if (!item) {
        item = { symbol: baseAsset, exchanges: new Set(), instrumentCount: 0, maxLeverage: 0.0 };
        stats.set(baseAsset, item);
      }
      item.exchanges.add(exchangeName);
      item.instrumentCount += 1;
      item.maxLeverage = Math.max(item.maxLeverage, toFloat(pair.max_leverage, 0.0) ?? 0.0);
    }
  }

  for (const [symbol, item] of stats) {
    if (item.exchanges.size < minExchangeCount) {
      stats.delete(symbol);
    }
  }
  return stats;
}

/** Exported for unit testing. */
export function rankCoinglassCandidates(
  candidateStats: Map<string, CandidateStats>,
  coreSymbols: string[],
  limit: number,
): string[] {
  const ranked = [...candidateStats.keys()].sort((a, b) => {
    const statsA = candidateStats.get(a) as CandidateStats;
    const statsB = candidateStats.get(b) as CandidateStats;
    if (statsA.exchanges.size !== statsB.exchanges.size) {
      return statsB.exchanges.size - statsA.exchanges.size;
    }
    if (statsA.instrumentCount !== statsB.instrumentCount) {
      return statsB.instrumentCount - statsA.instrumentCount;
    }
    if (statsA.maxLeverage !== statsB.maxLeverage) {
      return statsB.maxLeverage - statsA.maxLeverage;
    }
    if (a === b) {
      return 0;
    }
    return a > b ? -1 : 1;
  });

  const ordered: string[] = [];
  for (const symbol of [...coreSymbols, ...ranked]) {
    if (candidateStats.has(symbol) && !ordered.includes(symbol)) {
      ordered.push(symbol);
    }
    if (ordered.length >= limit) {
      break;
    }
  }
  return ordered;
}

/** Exported for unit testing. */
export function aggregateCoinglassPairs(
  pairs: CoinGlassPair[],
  exchanges: Set<string>,
  symbolStats: CandidateStats | null,
  quoteAsset: string,
): Row | null {
  const filtered = pairs.filter(
    (pair) =>
      (exchanges.size === 0 || exchanges.has(String(pair.exchange_name ?? ''))) &&
      pairSymbolMatchesQuote(pair, quoteAsset),
  );
  if (filtered.length === 0) {
    return null;
  }

  const primary = filtered.reduce((best, pair) =>
    (toFloat(pair.volume_usd, 0.0) ?? 0.0) > (toFloat(best.volume_usd, 0.0) ?? 0.0) ? pair : best,
  );
  const symbol = baseFromPair(primary, quoteAsset);
  const totalVolume = sumField(filtered, 'volume_usd');
  const totalOi = sumField(filtered, 'open_interest_usd');
  const longVolume = sumField(filtered, 'long_volume_usd');
  const shortVolume = sumField(filtered, 'short_volume_usd');
  const longLiq = sumField(filtered, 'long_liquidation_usd_24h');
  const shortLiq = sumField(filtered, 'short_liquidation_usd_24h');
  const funding = weightedAverage(filtered, 'funding_rate', 'open_interest_usd');

  return {
    symbol,
    contract_symbol: primary.instrument_id || `${symbol}${quoteAsset}`,
    base_asset: symbol,
    quote_asset: quoteAsset,
    primary_exchange: primary.exchange_name ?? null,
    data_source: 'coinglass',
    price_usd: weightedAverage(filtered, 'current_price', 'volume_usd'),
    index_price: weightedAverage(filtered, 'index_price', 'volume_usd'),
    price_change_24h_pct: weightedAverage(filtered, 'price_change_percent_24h', 'volume_usd'),
    quote_volume_usd: totalVolume,
    volume_change_percent_24h: weightedAverage(
      filtered,
      'volume_usd_change_percent_24h',
      'volume_usd',
    ),
    open_interest_usd: totalOi,
    oi_change_24h_pct: weightedAverage(
      filtered,
      'open_interest_change_percent_24h',
      'open_interest_usd',
    ),
    funding_rate_pct: funding,
    funding_annualized_pct: funding !== null ? fundingAnnualizedPct(funding / 100.0) : null,
    next_funding_time: primary.next_funding_time ?? null,
    long_volume_usd_24h: longVolume,
    short_volume_usd_24h: shortVolume,
    long_short_ratio: shortVolume > 0 ? longVolume / shortVolume : null,
    long_liquidation_usd_24h: longLiq,
    short_liquidation_usd_24h: shortLiq,
    open_interest_volume_ratio: totalVolume > 0 ? totalOi / totalVolume : null,
    coinglass_exchange_count: new Set(filtered.map((pair) => pair.exchange_name)).size,
    coinglass_instrument_count: symbolStats?.instrumentCount ?? null,
    coinglass_supported_exchange_count: symbolStats?.exchanges.size ?? 0,
  };
}

function normalizeCoingeckoGlobal(globalData: Record<string, unknown>): Record<string, unknown> {
  const totalMarketCap = asRecord(globalData.total_market_cap);
  const marketCapPercentage = asRecord(globalData.market_cap_percentage);
  return {
    total_market_cap_usd: toFloat(totalMarketCap.usd),
    market_cap_change_24h_pct: toFloat(globalData.market_cap_change_percentage_24h_usd),
    btc_dominance_pct: toFloat(marketCapPercentage.btc),
    eth_dominance_pct: toFloat(marketCapPercentage.eth),
    active_cryptocurrencies: globalData.active_cryptocurrencies ?? null,
    markets: globalData.markets ?? null,
  };
}

function normalizeCoingeckoCategories(
  categories: Record<string, unknown>[],
  limit: number,
): { leaders: Record<string, unknown>[]; laggards: Record<string, unknown>[] } {
  const normalized = categories
    .filter((item) => toFloat(item.market_cap_change_24h) !== null)
    .map((item) => ({
      id: item.id ?? null,
      name: item.name ?? null,
      market_cap_usd: toFloat(item.market_cap),
      market_cap_change_24h_pct: toFloat(item.market_cap_change_24h),
      volume_24h_usd: toFloat(item.volume_24h),
      top_3_coins: item.top_3_coins ?? [],
    }));

  const leaders = [...normalized]
    .sort((a, b) => (b.market_cap_change_24h_pct ?? 0) - (a.market_cap_change_24h_pct ?? 0))
    .slice(0, limit);
  const laggards = [...normalized]
    .sort((a, b) => (a.market_cap_change_24h_pct ?? 0) - (b.market_cap_change_24h_pct ?? 0))
    .slice(0, limit);
  return { leaders, laggards };
}

function sumField(rows: CoinGlassPair[], key: string): number {
  return rows.reduce((sum, row) => sum + (toFloat(row[key], 0.0) ?? 0.0), 0);
}
